crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1717 @@
+            morsel_seq: MorselSeq::default(),
+            metadata_rx,
+            use_statistics,
+            verbose,
+            reader_schema,
+            projection,
+            predicate,
+            slice_range: None, // Initialized later
+            memory_prefetch_func,
+            current_path_index: 0,
+            current_byte_source: Default::default(),
+            current_row_groups: Default::default(),
+            current_row_group_idx: 0,
+            current_max_row_group_height: 0,
+            current_row_offset: 0,
+            current_shared_file_state: Default::default(),
+        };
+
+        let row_group_decoder = self.init_row_group_decoder();
+        let row_group_decoder = Arc::new(row_group_decoder);
+
+        // Processes row group metadata and spawns I/O tasks to fetch row group data. This is
+        // currently spawned onto the CPU runtime as it does not directly make any async I/O calls,
+        // but instead it potentially performs predicate/slice evaluation on metadata. If we observe
+        // that under heavy CPU load scenarios the I/O throughput drops due to this task not being
+        // scheduled we can change it to be a high priority task.
+        let row_group_data_task_handle = async_executor::spawn(TaskPriority::Low, async move {
+            if start_rx.await.is_err() {
+                drop(row_group_data_fetcher);
+                return metadata_task_handle.await.unwrap();
+            }
+
+            if verbose {
+                eprintln!("[ParquetSource]: Starting row group data fetch")
+            }
+
+            // We must `recv()` from the `NotifyReceiver` before awaiting on the
+            // `normalized_slice_oneshot_rx`, as in the negative offset case the slice resolution
+            // only runs after the first notify.
+            if !row_group_data_fetcher.init_next_file_state().await {
+                drop(row_group_data_fetcher);
+                return metadata_task_handle.await.unwrap();
+            };
+
+            let slice_range = {
+                let Ok(slice) = normalized_slice_oneshot_rx.await else {
+                    // If we are here then the producer probably errored.
+                    drop(row_group_data_fetcher);
+                    return metadata_task_handle.await.unwrap();
+                };
+
+                slice.map(|(offset, len)| offset..offset + len)
+            };
+
+            row_group_data_fetcher.slice_range = slice_range;
+
+            // We can use buffer_unordered as the row group data contains the morsel seq.
+            let mut row_group_stream = row_group_data_fetcher
+                .into_stream()
+                .map(|x| async {
+                    match x {
+                        Ok(handle) => handle.await,
+                        Err(e) => Err(e),
+                    }
+                })
+                .buffer_unordered(row_group_prefetch_size)
```

Comment:
I've refactored to use a new approach for backpressure on concurrent downloading - the producer task now uses an internal buffer for in-flight download tasks. We no longer send task handles across the distributor channel, but instead we send downloaded data as they complete. Backpressure is now precisely determined by the size of this buffer instead of channel size, or permit-wrapping, which should be much cleaner.

Reply:
Nice :)